### PR TITLE
properly support lookahead expresions in begin matches

### PR DESCRIPTION
**Pull request by yyyc514**
- passes the FULL remaining buffer to `processLexeme` so
  that when it attemps to detemine the next mode it can
  run the ruleset against the FULL buffer, not just the
  terminator which resulted in a match (which will not
  be sufficient enough since lookaheads aren't included
  in the actual match)

### CODE REVIEW

**Comment by yyyc514** [comment]
I think we may need to do this same thing for `endOfMode` to support end matches, can anyone confirm?  The use case I cared about was begin matches, and that's what I've tested.

If this looks good there are a few syntaxes that can be updated now that we have PROPER lookahead support - since a few current have HACKS to deal with the prior broken support such as:

**XML.js**

```
        /*
        The lookahead pattern (?=...) ensures that 'begin' only matches
        '<style' as a single word, followed by a whitespace or an
        ending braket. The '$' is needed for the lexeme to be recognized
        by hljs.subMode() that tests lexemes outside the stream.
        */
        begin: '<style(?=\\s|>|$)', end: '>',
```

**Comment by yyyc514** [comment]
Can someone tel me the correct way to run tests from console?   `npm test` seems to be confused about where `../../build` is supposed to be.

**Comment by yyyc514** [comment]
Ha, now the issue is we have "broken"? rulesets that will actually match over and over if they can see the full data, such as Fortran:

```js
/(?=\b|\+|\-|\.)(?=\.\d|\d)(?:\d+)?(?:\.?\d*)(?:[de][+-]?\d+)?\b\.?/im
```

This can actually return a 0 length match, which won't push the position forward, then it'll find AGAIN the same 0 length match in an infinite loop. :-)

I'm thinking perhaps we should ignore 0 length matches.

**Comment by egor-rogov** [comment]
Hey @yyyc514, how is it going with this PR? Haven't heard from you in a while.

**Comment by marcoscaceres** [comment]
Heh, so this is how it all started 😂

**Comment by yyyc514** [comment]
I got a little distracted, it's true.  But I haven't forgotten this. :-)

**Comment by yyyc514** [comment]
Diff gets really confused around line 490 or so... probably easier to read lines 539 to 644 outside the diff.  They should be fairly easy to follow.

I shrunk `processLexeme` and moved a lot of the function into `doBeginMatch` and `doEndMatch`.

The end of `compileMode` moves into `buildModeRegex` which builds a 'super matcher' or whatever you'd like to call it. :-)  It pretends to be a regex (supporting `exec`) but also returns metadata reporting WHICH mode matched, etc... so we don't have to recheck.

**Comment by yyyc514** [comment]
Broke the browser build somehow, no idea how.

**Comment by egor-rogov** [comment]
I'm gonna look into this, but do not expect a quick feedback (:

**Comment by yyyc514** [comment]
No problem.  Nothing too advanced here I don’t think.  It’s really just a purer implementation of the goals of the original system.  Maybe it didn’t occur to them at the time you could just use capture groups to track which rule matched without the need to re-run the rule sets.

**Comment by yyyc514** [comment]
Not saying it's worth the time but so many places where we use returnBegin right now... probably 80% of them could be replaced with look-aheads with no negative issues, and perhaps even resulting in simplifying the grammar, like in the abnf case. (since most of the time you're replacing nesting with just a single, simple regex)

**Comment by egor-rogov** [comment]
I wish I had more time to grok this...
Anyway, what I see:
* all tests are passing
* speed remains roughly the same (around 9 sec on my laptop for `npm test`)
* lookaheads work!

So I can see no reason to delay this change. However I suggest to split it in 2 (maybe more) PRs: one for `highlight.js` itself, and the other(s) for improving grammar of languages.

@marcoscaceres You may want to have a look either, 'cause it changes the core of the system.

**Comment by egor-rogov** [comment]
> I think we may need to do this same thing for endOfMode to support end matches, can anyone confirm? The use case I cared about was begin matches, and that's what I've tested.

I think it makes perfect sense.
@yyyc514 it would be great if you can add a section to the developer's docs on what regexp features are supported (like lookaheads (thanks to this PR) and backreferences (thanks to  #1897)) and what are not (lookbehinds?).

**Comment by yyyc514** [comment]
Well, look behinds technically aren't in JS yet, are they?... but when they are added this PR should allow them to work just fine with begin matchers, and we should keep that in mind for end matchers also.

When you say "developer docs" which docs are you referring to exactly?

**Comment by yyyc514** [comment]
> However I suggest to split it in 2 (maybe more) PRs: one for highlight.js itself, and the other(s) for improving grammar of languages.

I can definitely do that.

**Comment by egor-rogov** [comment]
> When you say "developer docs" which docs are you referring to exactly?

The one in `docs/` folder (which goes to https://highlightjs.readthedocs.io/en/latest/)

**Comment by yyyc514** [comment]
Would you suggest a new document to talk about regex features?

**Comment by egor-rogov** [comment]
`language-guide` looks like a proper place for it.

**Comment by yyyc514** [comment]
I think I might keep it small and focus on what we DON'T support... I think the idea is its regex... it should just work... If someone wants to learn about regex there are all sorts of resources.

So right now the caveats (after this PR is merged):

- "look-ahead" doesn't work properly in end expressions.
- "look-behind" (when JS supports it) also doesn't work properly in end expressions.

**Comment by egor-rogov** [comment]
I thought that what we do support (and since what version) may be also important for those who previously stumbled on something unsupported.

**Comment by yyyc514** [comment]
Does that addition to docs help?

File: docs/language-guide.rst [hunk 1]
```diff
@@ -258,6 +258,23 @@ Many languages share common modes and regular expressions. Such expressions are
 at the end under "Common regexps" and "Common modes" titles. Use them when possible.
 
 
+Regular Expression Features
+---------------------------
+
+The goal of Highlight.js is to support whatever regex features Javascript itself supports.  IE, you're using real regular expressions, use them responsibly.  That said, due to the design of the parser, there are some caveats.  These are addressed below.
```
[the remark this answers is not in the document]
Comment (yyyc514) [reply docs/language-guide.rst]: That is what?

[the remark this answers is not in the document]
Comment (yyyc514) [reply docs/language-guide.rst]: OH LOL.... stupid acronyms...

**Comment by egor-rogov** [comment]
Ok, here we go!
I think we shouldn’t squash to keep separate commits.

**Pull request merged**